cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -2593,28 +2610,61 @@ int getPodIndexFromPodName(String podName)  {
+                return podOperations.listAsync(namespace, kafkaCluster.getSelectorLabels())
+                        .compose(pods -> {
+                            for (Pod pod : pods) {
+                                if (!needsRestartBecauseAddedOrRemovedJbodVolumes(pod, jbodStorage, kafkaCurrentReplicas, kafkaCluster.getReplicas()).isEmpty())   {
+                                    // At least one broker needs rolling update => we can trigger it without checking the other brokers
+                                    log.debug("{}: Kafka brokers needs rolling update to add or remove JBOD volumes", reconciliation);
+
+                                    return kafkaSetOperations.getAsync(namespace, KafkaCluster.kafkaClusterName(name))
+                                            .compose(sts -> {
+                                                if (sts != null) {
+                                                    int lastPodIndex = Math.min(kafkaCurrentReplicas, kafkaCluster.getReplicas()) - 1;
+                                                    return maybeRollKafkaInSequence(sts, podToCheck -> needsRestartBecauseAddedOrRemovedJbodVolumes(podToCheck, jbodStorage, kafkaCurrentReplicas, kafkaCluster.getReplicas()), 0, lastPodIndex);
+                                                } else {
+                                                    // STS does not exist => nothing to roll
+                                                    return withVoid(Future.succeededFuture());
+                                                }
+                                            });
+                                }
                             }
+
+                            log.debug("{}: No rolling update of Kafka brokers due to added or removed JBOD volumes is needed", reconciliation);
+                            return withVoid(Future.succeededFuture());
                         });
             } else {
                 return withVoid(Future.succeededFuture());
             }
         }
 
+        /**
+         * Checks whether the particular pod needs restart because of added or removed JBOD volumes. This method returns
+         * a list instead of boolean so that it can be used directly in the rolling update procedure and log proper
+         * reasons for the restart.
+         *
+         * @param pod               Broker pod which should be considered for restart
+         * @param desiredStorage    Desired storage configuration
+         * @param currentReplicas   Number of Kafka replicas before this reconciliation
+         * @param desiredReplicas   Number of desired Kafka replicas
+         *
+         * @return  Empty list when no restart is needed. List containing the restart reason if the restart is needed.
+         */
         private List<String> needsRestartBecauseAddedOrRemovedJbodVolumes(Pod pod, JbodStorage desiredStorage, int currentReplicas, int desiredReplicas)  {
```

Comment:
This method either needs to send an empty list or need a null check where it is called. Running this patch locally I ran into NPE.

Reply:
Then please provide more information about how to reproduce it or explain where does it return null.